src/cfn-mcp-server/awslabs/cfn_mcp_server/server.py

```diff
@@ -422,6 +426,108 @@ async def create_template(
+    tag_value: str | None = Field(
+        default=None,
+        description='Filter resources by tag value (requires tag_key to be specified)',
+    ),
+    limit: int = Field(
+        default=100,
+        description='Maximum number of resources to return (default: 100, max: 100 due to AWS API limits)',
+    ),
+    next_token: str | None = Field(
+        default=None, description='Pagination token from previous response to get next page'
+    ),
+    region: str | None = Field(
+        description='The AWS region that the operation should be performed in', default=None
+    ),
+) -> dict:
+    """List AWS resources with filtering support.
+
+    To preserve tokens usage, this tool allows you to filter resources by type resourcetypeprefix,
+    tag key, and tag value. It returns a paginated list of resource identifiers. It is recommended
+    to use the filtering parameters to reduce the number of resources returned,
+
+    This tool uses AWS CloudFormation's resource scan API with server-side filtering.
+    All parameters are optional except for the ResourceScanId (automatically handled).
+
+    Parameters:
+        resource_identifier: Filter by specific resource identifier (optional)
+        resource_type_prefix: Filter by resource type prefix (optional)
+        tag_key: Filter resources by tag key (optional)
+        tag_value: Filter resources by tag value (optional, requires tag_key)
+        limit: Maximum number of resources to return (1-100, AWS API limit)
+        next_token: AWS pagination token from previous response (optional)
+        region: AWS region to use (optional)
+
+    Returns:
+        Resource identifiers with filtering applied at the AWS API level and pagination metadata
+    """
+    return await list_resources_by_filter_impl(
+        resource_identifier=resource_identifier,
+        resource_type_prefix=resource_type_prefix,
+        tag_key=tag_key,
+        tag_value=tag_value,
+        limit=limit,
+        next_token=next_token,
+        region=region,
+    )
+
+
+@mcp.tool()
+async def list_related_resources(
+    resources: list = Field(
+        description='List of resources to find related resources for. Each resource should have resource_type and resource_identifier keys.'
+    ),
+    max_results: int = Field(
+        default=100,
+        description='Maximum number of related resources to return (default: 100, max: 100 due to AWS API limits)',
+    ),
+    next_token: str | None = Field(
+        default=None, description='Pagination token from previous response to get next page'
+    ),
+    region: str | None = Field(
+        description='The AWS region that the operation should be performed in', default=None
+    ),
+) -> dict:
+    """List AWS resources related to the specified resources.
+
+    This tool uses AWS CloudFormation's list_resource_scan_related_resources API
+    to find resources that are related to the specified input resources.
```

Comment:
you may want to add something like `Call with one resource at a time if you want to maintain explicit relationships`. The API doesn't return which related resource is related to which input resource.